### PR TITLE
#113 RsWithHeaders must reuse RsWithHeader

### PR DESCRIPTION
**Pull request by dmzaytsev**
For #113

Got rid code duplication in RsWithHeader and RsWithHeaders according the task description

### CODE REVIEW

**Comment by davvd** [comment]
@dmzaytsev Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@pinaf it's yours,please review

File: src/test/java/org/takes/rs/RsWithHeadersTest.java [hunk 1]
```diff
@@ -47,8 +48,8 @@ public void addsHeadersToResponse() throws IOException {
             new RsPrint(
                 new RsWithHeaders(
                     new RsEmpty(),
-                    "Host: www.example.com ",
-                    "Content-Type: text/xml "
+                    "Host: www.example.com",
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsWithHeadersTest.java]: @dmzaytsev let's extract this into a private final String variable.

File: src/test/java/org/takes/rs/RsWithHeadersTest.java [hunk 2]
```diff
@@ -47,8 +48,8 @@ public void addsHeadersToResponse() throws IOException {
...
+                    "Content-Type: text/xml"
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsWithHeadersTest.java]: @dmzaytsev let's extract this into a private final String variable.

File: src/test/java/org/takes/rs/RsWithHeadersTest.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
     /**
      * RsWithHeaders can add headers.
      * @throws IOException If some problem inside
+     * @checkstyle MultipleStringLiteralsCheck (20 lines)
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsWithHeadersTest.java]: @dmzaytsev and get rid of this also.

**Comment by pinaf** [comment]
@dmzaytsev nice :) 3 minor comments above.

**Comment by dmzaytsev** [comment]
@pinaf done :)

**Comment by pinaf** [comment]
@dmzaytsev thanks. this design is very "russian doll" style.

**Comment by dmzaytsev** [comment]
@pinaf how would you solved this problem?

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@dmzaytsev I wouldn't reuse the `RqWithHeader` class. Instead I'd build the headers once in the constructor as a list and just return it through the `head` method. It would be more efficient and, since the code is so simple, no readability would be lost.

**Comment by dmzaytsev** [comment]
@pinaf build failed .. weird
I had no errors before push

**Comment by dmzaytsev** [comment]
@pinaf use of the `RqWithHeader` was a condition of the task
anyway thanks for review

**Comment by pinaf** [comment]
@dmzaytsev only with java 8 on travis. seems like a network issue.
yes, you did the task as requested :)

**Comment by dmzaytsev** [comment]
@yegor256 take a look this PR #210 please.
Thank you

**Comment by pinaf** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@yegor256 this PR needs to be merged
We are waiting 5 days already.
Please pay your attention.

**Comment by dmzaytsev** [comment]
@yegor256 take a look this PR please

**Comment by yegor256** [comment]
@dmzaytsev this is a wrong design. I disagree with @pinaf here and even wrote an article about it. It will be published in a few days, but you can check the draft here: https://raw.githubusercontent.com/yegor256/blog/master/_drafts/ctors.md

Constructors should be light and never (ideally) do anything except assignments.

**Comment by pinaf** [comment]
@yegor256 what do you mean exactly? I pointed out an opinion but his design was different.

**Comment by pinaf** [comment]
@yegor256 full composition with one object instance per element in iterable?

**Comment by yegor256** [comment]
@pinaf I'm talking about this comment you made: https://github.com/yegor256/takes/pull/210#issuecomment-96238084

**Comment by pinaf** [comment]
@yegor256 yeah, I see - I agree with the article. However, you do note he chose the compositional style - kind of - instead of going with my opinion.

**Comment by yegor256** [comment]
@pinaf current implementation using `RsWithHeader` correctly, making many instances of this class. but this shouldn't happen in ctor, but in actual `head()` call instead

**Comment by pinaf** [comment]
@yegor256 ok - so the missing part is the delaying of the iteration, like you point out in the article. thanks!

**Comment by dmzaytsev** [comment]
@yegor256 useful article, thanks!
the logic has been moved to the `head()` call.
looks good now ?

File: src/main/java/org/takes/rs/RsWithHeaders.java [hunk 4]
```diff
@@ -80,4 +73,24 @@ public InputStream body() throws IOException {
...
+     * @throws IOException If fails
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    private static List<String> extend(final Response res,
+        final Iterable<? extends CharSequence> headers) throws IOException {
+        Response resp = res;
+        for (final CharSequence hdr: headers) {
+            resp = new RsWithHeader(resp, hdr);
+        }
+        final List<String> ret = new ArrayList<String>(1);
```
Comment (yegor256) [inline src/main/java/org/takes/rs/RsWithHeaders.java]: I didn't get this... why can't you just do `return resp.head()`?

**Comment by yegor256** [comment]
@dmzaytsev yes, better, but you over-engineered it, I believe. see my comment

**Comment by dmzaytsev** [comment]
@yegor256 you're right
take a look please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by yegor256** [comment]
@dmzaytsev @pinaf here is the article published: http://www.yegor256.com/2015/05/07/ctors-must-be-code-free.html

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks for the review and a good remark

**Comment by davvd** [comment]
@pinaf I added 10 mins to @elenavolokhova (for QA review) in transaction 56872197... Thanks so much! Your account was topped up for **25 mins** (transaction ID is `AP-2NS64077DN239632Y`, the task took 264 hours and 25 mins)... you're getting extra minutes here (c=10)... +25 added to your rating, current score is: [+7693](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor deploy now pls

**Pull request merged**